### PR TITLE
Add the ability to add individual files exefs with mod loader

### PR DESCRIPTION
**Pull request by shadowninja108**

### CODE REVIEW

File: Ryujinx.HLE/HOS/ApplicationLoader.cs [hunk 1]
```diff
@@ -463,45 +480,48 @@ private void LoadExeFs(IFileSystem codeFs, Npdm metaData = null)
...
-                foreach (DirectoryEntryEx file in codeFs.EnumerateEntries("/", exePrefix))
-                {
-                    if (Path.GetExtension(file.Name) != string.Empty)
-                    {
-                        continue;
-                    }
+                string name = ExeFsPrefixes[i];
 
-                    Logger.Info?.Print(LogClass.Loader, $"Loading {file.Name}...");
+                if (!codeFs.FileExists(name))
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/ApplicationLoader.cs:489]: Add `{}`

File: Ryujinx.HLE/HOS/ApplicationLoader.cs [hunk 2]
```diff
@@ -463,45 +480,48 @@ private void LoadExeFs(IFileSystem codeFs, Npdm metaData = null)
...
 
-                    codeFs.OpenFile(out IFile nsoFile, file.FullPath.ToU8Span(), OpenMode.Read).ThrowIfFailure();
+                Logger.Info?.Print(LogClass.Loader, $"Loading {name}...");
 
-                    NsoExecutable nso = new NsoExecutable(nsoFile.AsStorage(), file.Name);
+                codeFs.OpenFile(out IFile nsoFile, $"/{name}".ToU8Span(), OpenMode.Read).ThrowIfFailure();
+
+                NsoExecutable nso = new NsoExecutable(nsoFile.AsStorage(), name);
+
+                nsos[i] = nso;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/ApplicationLoader.cs:499]: Remove extra empty line

File: Ryujinx.HLE/HOS/ApplicationLoader.cs [hunk 3]
```diff
@@ -463,45 +480,48 @@ private void LoadExeFs(IFileSystem codeFs, Npdm metaData = null)
...
             // ExeFs file replacements
-            bool modified = _fileSystem.ModLoader.ApplyExefsMods(TitleId, nsos);
+            ModLoadResult modres = _fileSystem.ModLoader.ApplyExefsMods(TitleId, nsos);
+
+            // collect the nsos, ignoring ones that aren't used
+            NsoExecutable[] programs = nsos.Where(x => x != null).ToArray();
 
-            NsoExecutable[] programs = nsos.ToArray();
+            // take the npdm from mods if present
+            if(modres.Npdm != null)
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/ApplicationLoader.cs:509]: Add `{}`

**Review by AcK77 — CHANGES_REQUESTED** [review]

**Comment by jduncanator** [comment]
Please add a description describing the changes you've made, and how they work.

File: Ryujinx.HLE/HOS/ApplicationLoader.cs [hunk 4]
```diff
@@ -30,7 +31,23 @@ namespace Ryujinx.HLE.HOS
...
+            "subsdk3",
+            "subsdk4",
+            "subsdk5",
+            "subsdk6",
+            "subsdk7",
+            "subsdk8",
+            "subsdk9",
+            "sdk"
+        };
+
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/ApplicationLoader.cs:50]: Remove extra empty line

File: Ryujinx.HLE/HOS/ApplicationLoader.cs [hunk 5]
```diff
@@ -463,45 +480,49 @@ private void LoadExeFs(IFileSystem codeFs, Npdm metaData = null)
...
+                codeFs.OpenFile(out IFile nsoFile, $"/{name}".ToU8Span(), OpenMode.Read).ThrowIfFailure();
 
-                    nsos.Add(nso);
-                }
+                nsos[i] = new NsoExecutable(nsoFile.AsStorage(), name);
             }
 
             // ExeFs file replacements
-            bool modified = _fileSystem.ModLoader.ApplyExefsMods(TitleId, nsos);
+            ModLoadResult modres = _fileSystem.ModLoader.ApplyExefsMods(TitleId, nsos);
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/ApplicationLoader.cs:502]: What about :
```suggestion
            ModLoadResult modLoadResult = _fileSystem.ModLoader.ApplyExefsMods(TitleId, nsos);
```

File: Ryujinx.HLE/HOS/ModLoader.cs [hunk 6]
```diff
@@ -381,66 +382,82 @@ internal bool ReplaceExefsPartition(ulong titleId, ref IFileSystem exefs)
...
+            public BitVector32 Stubs;
+            public BitVector32 Replaces;
+            public Npdm Npdm;
+
+            public bool Modified => (Stubs.Data | Replaces.Data) != 0;
+        }
+
+        internal ModLoadResult ApplyExefsMods(ulong titleId, NsoExecutable[] nsos)
+        {
+            ModLoadResult res = new ModLoadResult();
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/ModLoader.cs:396]: You can set the fields like that. And probably rename the var to something a bit more explicit.
```suggestion
            ModLoadResult modLoadResult = new ModLoadResult
            {
                Stubs    = new BitVector32(),
                Replaces = new BitVector32()
            };
```

File: Ryujinx.HLE/HOS/ModLoader.cs [hunk 7]
```diff
@@ -381,66 +382,82 @@ internal bool ReplaceExefsPartition(ulong titleId, ref IFileSystem exefs)
...
-                        replaced = true;
+                    res.Stubs[1 << i] |= File.Exists(Path.Combine(mod.Path.FullName, nsoName + StubExtension));
+                }
 
+                FileInfo npdmFile = new FileInfo(Path.Combine(mod.Path.FullName, "main.npdm"));
+                if(npdmFile.Exists)
+                {
+                    if(res.Npdm != null)
+                    {
+                        Logger.Warning?.Print(LogClass.ModLoader, "Multiple replacements to 'main.npdm'");
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/ModLoader.cs:442]: Add a new empty line after

File: Ryujinx.HLE/HOS/ModLoader.cs [hunk 8]
```diff
@@ -381,66 +382,82 @@ internal bool ReplaceExefsPartition(ulong titleId, ref IFileSystem exefs)
...
                         continue;
                     }
 
-                    stubs[1 << i] |= File.Exists(Path.Combine(mod.Path.FullName, nsoName + StubExtension));
+                    Logger.Info?.Print(LogClass.ModLoader, $"main.npdm replaced");
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/ModLoader.cs:446]: If the NPDM is "replaced", you should put this after the next line.

File: Ryujinx.HLE/HOS/ModLoader.cs [hunk 9]
```diff
@@ -381,66 +382,86 @@ internal bool ReplaceExefsPartition(ulong titleId, ref IFileSystem exefs)
...
+                {
+                    if(modLoadResult.Npdm != null)
+                    {
+                        Logger.Warning?.Print(LogClass.ModLoader, "Multiple replacements to 'main.npdm'");
 
                         continue;
                     }
 
-                    stubs[1 << i] |= File.Exists(Path.Combine(mod.Path.FullName, nsoName + StubExtension));
+                    modLoadResult.Npdm = new Npdm(npdmFile.OpenRead());
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/ModLoader.cs:450]:
```suggestion
                    modLoadResult.Npdm = new Npdm(npdmFile.OpenRead());
                    
```

**Review by AcK77 — APPROVED** [review]
Lgtm! Thanks!

**Review by Thog — APPROVED** [review]

**Comment by gdkchan** [comment]
Those changes broke loading unpacked ROMs. They now always fail with `Process initialization returned error "InvalidMemRange".`.

File: Ryujinx.HLE/HOS/ModLoader.cs [hunk 10]
```diff
@@ -381,66 +382,87 @@ internal bool ReplaceExefsPartition(ulong titleId, ref IFileSystem exefs)
...
                         nsos[i] = new NsoExecutable(nsoFile.OpenRead().AsStorage(), nsoName);
                         Logger.Info?.Print(LogClass.ModLoader, $"NSO '{nsoName}' replaced");
+                    }
 
-                        replaced = true;
+                    modLoadResult.Stubs[1 << i] |= File.Exists(Path.Combine(mod.Path.FullName, nsoName + StubExtension));
+                }
+
+                FileInfo npdmFile = new FileInfo(Path.Combine(mod.Path.FullName, "main.npdm"));
+                if(npdmFile.Exists)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/ModLoader.cs:441]: `if (npdmFile.Exists)`.

File: Ryujinx.HLE/HOS/ModLoader.cs [hunk 11]
```diff
@@ -381,66 +382,87 @@ internal bool ReplaceExefsPartition(ulong titleId, ref IFileSystem exefs)
...
+                {
+                    if(modLoadResult.Npdm != null)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/ModLoader.cs:443]: `if (modLoadResult.Npdm != null)`.

File: Ryujinx.HLE/HOS/ApplicationLoader.cs [hunk 12]
```diff
@@ -463,45 +479,56 @@ private void LoadExeFs(IFileSystem codeFs, Npdm metaData = null)
 
             metaData ??= ReadNpdm(codeFs);
 
-            List<NsoExecutable> nsos = new List<NsoExecutable>();
+            NsoExecutable[] nsos = new NsoExecutable[ExeFsPrefixes.Length];
 
-            foreach (string exePrefix in ExeFsPrefixes) // Load binaries with standard prefixes
+            for(int i = 0; i < nsos.Length; i++)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/ApplicationLoader.cs:484]: `for (int i = 0; i < nsos.Length; i++)`.

**Pull request merged**